crates/mun_runtime/src/test.rs

```diff
@@ -1,95 +1,143 @@
 use crate::{Runtime, RuntimeBuilder};
-use mun_compiler::CompilerOptions;
+use mun_abi::Reflection;
+use mun_compiler::{ColorChoice, Config, Driver, FileId, PathOrInline, RelativePathBuf};
 use std::path::PathBuf;
+use std::thread::sleep;
+use std::time::Duration;
 
-struct CompileResult {
+/// Implements a compiler and runtime in one that can invoke functions. Use of the TestDriver
+/// enables quick testing of Mun constructs in the runtime with hot-reloading support.
+struct TestDriver {
     _temp_dir: tempfile::TempDir,
-    result: PathBuf,
+    out_path: PathBuf,
+    file_id: FileId,
+    driver: Driver,
+    runtime: Runtime,
 }
 
-impl CompileResult {
-    /// Construct a runtime from the compilation result that can be used to execute the compiled
-    /// files.
-    pub fn new_runtime(&self) -> Runtime {
-        RuntimeBuilder::new(&self.result).spawn().unwrap()
+impl TestDriver {
+    /// Construct a new TestDriver from a single Mun source
+    fn new(text: &str) -> Self {
+        let temp_dir = tempfile::TempDir::new().unwrap();
+        let config = Config {
+            out_dir: Some(temp_dir.path().to_path_buf()),
+            ..Config::default()
+        };
+        let input = PathOrInline::Inline {
+            rel_path: RelativePathBuf::from("main.mun"),
+            contents: text.to_owned(),
+        };
+        let (driver, file_id) = Driver::with_file(config, input).unwrap();
+        let mut err_stream = mun_compiler::StandardStream::stderr(ColorChoice::Auto);
+        if driver.emit_diagnostics(&mut err_stream).unwrap() {
+            panic!("compiler errors..")
+        }
+        let out_path = driver.write_assembly(file_id).unwrap().unwrap();
+        let runtime = RuntimeBuilder::new(&out_path).spawn().unwrap();
+        TestDriver {
+            _temp_dir: temp_dir,
+            driver,
+            out_path,
+            file_id,
+            runtime,
+        }
+    }
+
+    /// Updates the text of the Mun source and ensure the generated assembly has been reloaded.
+    fn update(&mut self, text: &str) {
+        self.driver.set_file_text(self.file_id, text);
+        let out_path = self.driver.write_assembly(self.file_id).unwrap().unwrap();
+        assert_eq!(
+            &out_path, &self.out_path,
+            "recompiling did not result in the same assembly"
+        );
+        let start_time = std::time::Instant::now();
+        while !self.runtime.update() {
+            let now = std::time::Instant::now();
+            if now - start_time > std::time::Duration::from_secs(10) {
+                panic!("runtime did not update after recompilation within 10secs");
+            } else {
+                sleep(Duration::from_millis(1));
+            }
+        }
+    }
+
+    /// Calls the function with the specified name
+    fn invoke0<T: Reflection>(&mut self, fn_name: &str) -> T {
```

Comment:
Why don't you add a `runtime_mut()` function to `TestDriver`? That way you don't have to duplicate the `invoke_fn` logic.

Reply:
Because this is shorter which I find much clearer.

Reply:
The trade-off is that a lot more code is added overall.